site/learn/expert-advice/authentication/webauthn-explained.md

```diff
@@ -0,0 +1,478 @@
+-The private key is kept on the authenticator
+-The public key is passed up to the website's server via the browser
+
+This allows the website's server to store the credential and associate it with the user. When an authentication ceremony happens, the website will therefore have the user's public key ready to go. 
+
+Registration is tied to a specific website. If you use the same authenticator on multiple websites, each website will have its own corresponding key pair.
+
+{% include _image.liquid src="/assets/img/advice/webauthn-explained/keys-per-website.png" alt="Each website has its own keypair." class="img-fluid" figure=false %}
+
+Let's look at the code flow for the registration ceremony. The options are a bit more complicated than they are for the authentication ceremony. 
+
+```javascript
+options =  {
+  publicKey: {
+    rp: { id: "cosmosclownstore.com", 
+          name: "Cosmo’s Clown Store" },
+    user: { id: "1234", 
+            name: "krusty", 
+            displayName: "krusty The Clown" },
+    challenge: "...",
+    pubKeyCredParams: [ { type: "public-key", alg: -7 }]
+    //...
+  }
+}
+```
+
+You again have the `publicKey` as the top level options field, indicating that you are using WebAuthn.
+
+Then there is the `rp` object, which has information about the website that is requesting the registration operation. This includes the name to be displayed to the end user by the authenticator.
+
+You also have the `user` field, which is information used to tie the user to the credential. Additionally, to create the key, you need to specify the algorithm to use. Do this by passing the correct parameters in `pubKeyCredParams`. Above, the value is `-7`. This is an item in the [IANA COSE Algorithms registry](https://www.iana.org/assignments/cose/cose.xhtml).
+
+After creating the options object, call the `create` method on the credentials API and if it succeeds, the user will be registered with cosmosclownstore.com.
+
+```javascript
+navigator.credentials.create(options)
+  .then(function (newCredentialInfo) {
+    // Send credential info to server 
+  }).catch(function (err) {
+    // No acceptable authenticator or user refused consent. 
+  });
+```
+
+Here's an example response.
+
+```javascript
+{
+  response: {
+    clientDataJSON: { 
+      type: “webauthn.create”,
+      challenge: ... // should match initial challenge
+      origin: ... //
+    },
+    attestationObject: {
+      authData: {
+        attestedCredentialData: {
+          credentialId: ..., 
+          credentialPublicKey: ...,
+        }
+      }, // other stuff about attestation 
+    }
+  }
+}
+```
+
+The `attestationObject` contains items such as:
+
+- the credential Id 
+- the public key 
+
+Store the `credentialId` to identify the user and the public key used when verifying the assertions generated by the other ceremonies. You must also verify the `challenge` and `type` in the `clientDataJSON` object.
```

Comment:
The `credentialId` doesn't identify the user directly. It should be stored and associated with the user that registered the credential.
- `credentialId` is used to request that particular credential from an available authenticator
- `credentialPublicKey` is used to validate the signature on an assertion